internal/mirror/mirror.go

```diff
@@ -0,0 +1,276 @@
+			curFiles, err := s.downloadIndex(mirror)
+			if err != nil {
+				logger.Errorw("unable to download index.txt", "error", err)
+			}
+
+			// figure out the actions
+			actions := make(map[string]*FileStatus, len(curFiles))
+			for _, kf := range knownFiles {
+				actions[kf.Filename] = &FileStatus{
+					MirrorFile: kf,
+				}
+			}
+			for i, curFile := range curFiles {
+				lastIdx := strings.LastIndex(curFile, "/")
+				fileName := curFile[lastIdx+1:]
+				if cur, ok := actions[fileName]; ok {
+					cur.DownloadPath = curFile
+					cur.Filename = fileName
+					cur.Order = i
+				} else {
+					actions[fileName] = &FileStatus{
+						Order:        i,
+						Filename:     fileName,
+						DownloadPath: curFile,
+					}
+				}
+			}
+
+			// delete any files we no longer have
+			for filename, status := range actions {
+				if status.needsDelete() {
+					logger.Infow("deleting stale export file", "file", filename)
+
+					bsCtx, cancel := context.WithTimeout(ctx, time.Second*30)
+					defer cancel()
+					// If there's a race and the file was already deleted, no error is returned.
+					if err := blobstore.DeleteObject(bsCtx, mirror.CloudStorageBucket, filename); err != nil {
+						logger.Errorw("unable to delete file", "bucket", mirror.CloudStorageBucket, "file", filename, "error", err)
+						continue
+					}
+					// Remove from map, so will be deleted from db.
+					delete(actions, filename)
+				}
+			}
+
+			// download any files we don't have
+			indexObjects := make([]*FileStatus, 0, len(actions))
+			// quickly bypass anything that doesn't need to be downloaded.
+			// This ensures that if we timeout, files we previously knew about remain saved.
+			for _, status := range actions {
+				if !status.needsDownload() {
+					status.Saved = true
+					indexObjects = append(indexObjects, status)
+				}
+			}
+			logger.Infow("mirror state", "mirrorID", mirror.ID, "total", len(actions), "existing", len(indexObjects), "todo", len(actions)-len(indexObjects))
+
+			// go through again and process downloads.
+			for filename, status := range actions {
+				if deadlinePassed(deadline) {
+					logger.Warnw("mirror didn't complete processing before timeout", "mirrorID", mirror.ID)
+					break
+				}
+
+				if status.needsDownload() {
+					logger.Infow("mirroring export file", "mirrorID", mirror.ID, "file", filename)
+
+					client := &http.Client{
+						Timeout: s.config.ExportFileDownloadTimeout,
+					}
+					resp, err := client.Get(status.DownloadPath)
+					if err != nil {
+						logger.Errorw("unable to download file", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					defer resp.Body.Close()
+					bytes, err := ioutil.ReadAll(resp.Body)
+					if err != nil {
+						logger.Errorw("unable to read file", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					// Write this file to the blob store.
+					bsCtx, cancel := context.WithTimeout(ctx, time.Minute)
```

Comment:
Why is the other 30*time.Second and this is one minute?

Reply:
one is delete and one is write